### PR TITLE
[SPARK-24244][SPARK-24368][SQL] Passing only required columns to the CSV parser

### PR DESCRIPTION
**Pull request by MaxGekk**
## What changes were proposed in this pull request?

uniVocity parser allows to specify only required column names or indexes for [parsing](https://www.univocity.com/pages/parsers-tutorial) like:

```
// Here we select only the columns by their indexes.
// The parser just skips the values in other columns
parserSettings.selectIndexes(4, 0, 1);
CsvParser parser = new CsvParser(parserSettings);
```
In this PR, I propose to extract indexes from required schema and pass them into the CSV parser. Benchmarks show the following improvements in parsing of 1000 columns:

```
Select 100 columns out of 1000: x1.76
Select 1 column out of 1000: x2
```

**Note**: Comparing to current implementation, the changes can return different result for malformed rows in the `DROPMALFORMED` and `FAILFAST` modes if only subset of all columns is requested. To have previous behavior, set `spark.sql.csv.parser.columnPruning.enabled` to `false`.
 
## How was this patch tested?

It was tested by new test which selects 3 columns out of 15, by existing tests and by new benchmarks.

### CODE REVIEW

**Comment by MaxGekk** [comment]
The difference between this PR and #21296 is that the `columnPruning` is passed to CSVOptions as a parameter. It should fix flaky `UnivocityParserSuite`.

**Comment by MaxGekk** [comment]
jenkins, retest this, please

**Comment by HyukjinKwon** [comment]
retest this please

**Comment by MaxGekk** [comment]
jenkins, retest this, please

**Comment by MaxGekk** [comment]
jenkins, retest this, please

**Comment by MaxGekk** [comment]
jenkins, retest this, please

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/csv/CSVOptions.scala [hunk 1]
```diff
@@ -29,17 +29,20 @@ import org.apache.spark.sql.catalyst.util._
...
+    defaultColumnNameOfCorruptRecord: String,
+    val columnPruning: Boolean)
   extends Logging with Serializable {
 
   def this(
     parameters: Map[String, String],
     defaultTimeZoneId: String,
-    defaultColumnNameOfCorruptRecord: String = "") = {
+    defaultColumnNameOfCorruptRecord: String = "",
+    columnPruning: Boolean = false) = {
```
Comment (gatorsmile) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/csv/CSVOptions.scala]: Let us do not set the default value for `columnPruning`? We might lose the pruning opportunity if we call this constructor.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/csv/CSVSuite.scala [hunk 2]
```diff
@@ -1383,4 +1385,31 @@ class CSVSuite extends QueryTest with SharedSQLContext with SQLTestUtils with Te
...
+
+      val odf = spark.createDataFrame(List(
+        Row(1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15),
+        Row(-1, -2, -3, -4, -5, -6, -7, -8, -9, -10, -11, -12, -13, -14, -15)
+      ).asJava, schema)
+      odf.write.csv(path.getCanonicalPath)
+      val idf = spark.read
+        .schema(schema)
+        .csv(path.getCanonicalPath)
+        .select('f15, 'f10, 'f5)
```
Comment (gatorsmile) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/csv/CSVSuite.scala]: Could you add an extreme test case? Try `count(1)` on csv files? That means zero column is required.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/csv/CSVBenchmarks.scala [hunk 3]
```diff
@@ -74,7 +74,49 @@ object CSVBenchmarks {
...
+      }
+
+      /*
+      Intel(R) Core(TM) i7-7920HQ CPU @ 3.10GHz
+
+      Wide rows with 1000 columns:         Best/Avg Time(ms)    Rate(M/s)   Per Row(ns)   Relative
+      --------------------------------------------------------------------------------------------
+      Select 1000 columns                     76910 / 78065          0.0       76909.8       1.0X
+      Select 100 columns                      28625 / 32884          0.0       28625.1       2.7X
+      Select one column                       22498 / 22669          0.0       22497.8       3.4X
```
Comment (gatorsmile) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/csv/CSVBenchmarks.scala]: count(1) too?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/csv/CSVOptions.scala (hunk 1 above)
Comment (MaxGekk) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/csv/CSVOptions.scala]: The constructor with disabled `columnPruning` is called in the CSV writer and 30 times from test suites like `UnivocityParserSuite` and `CSVInferSchemaSuite`.

> We might lose the pruning opportunity if we call this constructor.

ok. I will enable it by default.

Comment (gatorsmile) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/csv/CSVOptions.scala]: always enabling it is also not right. Can we remove the default?

File: sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/csv/CSVSuite.scala (hunk 2 above)
Comment (MaxGekk) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/csv/CSVSuite.scala]: added an assert for count(). In the `CSVSuite`, there are a few tests with `count()` over malformed csv files.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/csv/CSVOptions.scala (hunk 1 above)
Comment (MaxGekk) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/csv/CSVOptions.scala]: removed

File: sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/csv/CSVBenchmarks.scala (hunk 3 above)
Comment (MaxGekk) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/csv/CSVBenchmarks.scala]: sure, added `count()`

**Comment by gatorsmile** [comment]
LGTM

Thanks! Merged to master.

**Pull request closed without merging**